README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->



[![Travis-CI Build Status](https://travis-ci.org/hrbrmstr/censys.svg?branch=master)](https://travis-ci.org/hrbrmstr/censys) [![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/censys)](https://cran.r-project.org/package=censys)

`censys` is an R package interface to the [Censys API](https://censys.io/api)

Censys is a search engine that enables researchers to ask questions about the hosts and networks that compose the Internet. Censys collects data on hosts and websites through daily [ZMap](https://zmap.io/) and [ZGrab](https://github.com/zmap/zgrab) scans of the IPv4 address space, in turn maintaining a database of how hosts and websites are configured. Researchers can interact with this data through a search interface, report builder, and SQL engine.

[Censys tutorial](https://www.censys.io/tutorial).

The following functions are implemented:

- `censys_export_download`:	Download export job files to a specified directory
- `censys_export_job_status`:	Get status of a Censys export job
- `censys_get_job_result`:	Get results of completed Censys SQL query job
- `censys_get_job_status`:	Get status of a Censys SQL query job
- `censys_query`:	Issue SQL Queries against the Censys API
- `censys_report`:	Create aggregate reports on the breakdown of a field in the result set of a query
- `censys_search`:	Perform queries against Censys data
- `censys_series`:	List all series that can be queried from the SQL interface
- `censys_series_details`:	Get details about a series, including the list of tables and schema for the series
- `censys_start_export`:	Export large datasets and structured records from Censys to JSON or CSV files
- `get_series`:	Retrieve data on the types of scans Censys regularly performs ("series").
- `view_document`:	Retrieve data that Censys has about a specific host, website, or certificate.
- `view_result`:	Retrieve data on a particular scan "result"
- `view_series`:	Retrieve data that Censys has about a particular series

### Installation


```r
devtools::install_github("hrbrmstr/censys")
```



### Usage


```r
library(censys)

# current verison
packageVersion("censys")
#> [1] '0.1.0'

library(purrr)
library(dplyr)

res <- censys_query("
SELECT p80.http.get.headers.server, p80.http.get.headers.www_authenticate, location.country, autonomous_system.asn
FROM ipv4.20161206
WHERE REGEXP_MATCH(p80.http.get.headers.server, r'gen[56]')
")

inf <- censys_get_job_status(res$job_id)

Sys.sleep(10) # giving it some time to process for the README

map(1:3, ~censys_get_job_result(inf$job_id, .)) %>% 
  map(c("rows", "f")) %>% 
  flatten() %>% 
  map("v") %>% 
  map_df(~setNames(as.list(.), c("server", "auth", "geo", "asn"))) %>% 
  count(geo, sort=TRUE)
#> # A tibble: 54 × 2
#>               geo     n
#>             <chr> <int>
#> 1   United States   645
#> 2         Germany   338
#> 3         Vietnam   233
#> 4         Austria   135
#> 5          France   127
#> 6           Japan    77
#> 7  Czech Republic    64
#> 8           Italy    60
#> 9          Russia    57
#> 10         Canada    44
#> # ... with 44 more rows
```

### Test Results


```r
library(censys)
library(testthat)

date()
#> [1] "Sat Dec 31 10:14:03 2016"

test_dir("tests/")
#> testthat results ========================================================================================================
#> OK: 2 SKIPPED: 0 FAILED: 0
#> 
#> DONE ===================================================================================================================
```

### Code of Conduct

Please note that this project is released with a [Contributor Code of Conduct](CONDUCT.md). 
By participating in this project you agree to abide by its terms.


NEWS.md
# censys 0.1.0

- CRAN release
- full API working
- pkgdown for github.io site

# censys 0.1.0.9000

* 99% of Censys API represented
* Added a `NEWS.md` file to track changes to the package.





tests/testthat.R
library(testthat)
library(censys)

test_check("censys")


tests/testthat/test-censys.R
context("basic functionality")
test_that("we can do something", {

  testthat::skip_on_cran()
  testthat::skip_on_travis()

  expect_gte(length(get_series()), 2)

  q <- censys_query("SELECT p443.https.tls.cipher_suite.name, count(ip) FROM ipv4
                   WHERE p443.https.tls.validation.browser_trusted=true
                   GROUP BY p443.https.tls.cipher_suite.name;")

  expect_equal(sort(names(q)), c("configuration", "job_id", "status"))

})


R/data.r
#' Retrieve data on the types of scans Censys regularly performs ("series").
#'
#' You must have both \code{CENSYS_API_ID} and \code{CENSYS_API_SECRET} present in the
#' R environment for the functions in this package to work. It is highly suggested that
#' you place those in \code{~/.Renviron} at least for interactive work.
#'
#' @return list of series metadata
#' @note Censys ID & Secret must be in the R environment
#' @export
#' @references Census API: \url{https://www.censys.io/api/v1/docs/data}
#' @examples \dontrun{
#' scans <- get_series()
#' names(scans$raw_series)
#' names(scans$primary_series)
#' }
get_series <- function() {

  result <- httr::GET(CENSYS_API_URL %s+% "data", check_api())

  httr::stop_for_status(result)

  srs <- jsonlite::fromJSON(content(result, as="text"), flatten=TRUE)

  class(srs) <- c("censys_srs", class(srs))

  srs

}


#' Retrieve data that Censys has about a particular series
#'
#' A "series" is a scan of the same protocol and destination accross time,
#' including the list of scans.
#'
#' You must have both \code{CENSYS_API_ID} and \code{CENSYS_API_SECRET} present in the
#' R environment for the functions in this package to work. It is highly suggested that
#' you place those in \code{~/.Renviron} at least for interactive work.
#'
#' @param series_id Censys series id (e.g. \code{"22-ssh-banner-full_ipv4"})
#' @return list of specific series details
#' @export
#' @references Census API: \url{https://www.censys.io/api/v1/docs/data}
#' @examples \dontrun{
#' view_series("443-https-tls-full_ipv4")
#' }
view_series <- function(series_id) {

  result <- httr::GET(CENSYS_API_URL %s+% "data/" %s+% series_id, check_api())

  httr::stop_for_status(result)

  srs <- jsonlite::fromJSON(content(result, as="text"), flatten=TRUE)

  class(srs) <- c("censys_srs_info", class(srs))

  srs

}

#' Retrieve data on a particular scan "result"
#'
#' Generally used after a call to either \code{get_series} or \code{view_series}.
#'
#' You must have both \code{CENSYS_API_ID} and \code{CENSYS_API_SECRET} present in the
#' R environment for the functions in this package to work. It is highly suggested that
#' you place those in \code{~/.Renviron} at least for interactive work.
#'
#' @param series_id Censys series id (e.g. \code{"22-ssh-banner-full_ipv4"})
#' @param result_id Censys series result id (e.g. \code{"20150930T0056"})
#' @return list of specific series result details
#' @export
#' @references Census API: \url{https://www.censys.io/api/v1/docs/data}
view_result <- function(series_id, result_id) {

  result <- httr::GET(CENSYS_API_URL %s+% "data/" %s+% series_id %s+% "/" %s+% result_id,
                check_api())

  httr::stop_for_status(result)

  srs <- jsonlite::fromJSON(content(result, as="text"), flatten=TRUE)

  class(srs) <- c("censys_srs_res", class(srs))

  srs

}


R/view.r
#' Retrieve data that Censys has about a specific host, website, or certificate.
#'
#' The view endpoint fetches the structured data we have about a specific host, website,
#' or certificate once you know the host's IP address, website's domain, or certificate's
#' SHA-256 fingerprint.
#'
#' You must have both \code{CENSYS_API_ID} and \code{CENSYS_API_SECRET} present in the
#' R environment for the functions in this package to work. It is highly suggested that
#' you place those in \code{~/.Renviron} at least for interactive work.
#'
#' @param index The search index the document is in. Must be one of either
#'        \code{ipv4}, \code{websites}, or \code{certificates}.
#' @param id The ID of the document you are requesting. In the \code{ipv4} index,
#'        this is IP address (e.g., \code{192.168.1.1}), domain name in the
#'        \code{websites} index (e.g., \code{google.com}) and SHA-256 fingerprint
#'        in the \code{certificates} index (e.g.,
#'        \code{9d3b51a6b80daf76e074730f19dc01e643ca0c3127d8f48be64cf3302f6622cc}).
#' @return list of information about the endpoint
#' @export
#' @examples \dontrun{
#' view_document("google.com")
#' }
view_document <- function(index, id) {

  result <- httr::GET(CENSYS_API_URL %s+% "view/" %s+% index %s+% "/" %s+% id, check_api())

  httr::stop_for_status(result)

  srs <- jsonlite::fromJSON(content(result, as="text"), flatten=TRUE)

  class(srs) <- c("censys_view_doc", class(srs))

  srs

}


R/export.r
#' Export large datasets and structured records from Censys to JSON or CSV files
#'
#' The Export API allows exporting large datasets and structured records from Censys to
#' JSON or CSV files. Unlike the query endpoint, there are no limits on the type or amount
#' of data returned.
#'
#' Exports are executed as asynchronous jobs. You must first start a job. If the query is
#' parsed successfully, the call will return a job ID, which is used in subsequent calls
#' to get job. Once a job executes successfully, the get job endpoint will provide a list
#' of 128MB JSON files that are available for download for 24 hours. Jobs typically
#' require 15-30 seconds to execute.
#'
#' You must have both \code{CENSYS_API_ID} and \code{CENSYS_API_SECRET} present in the
#' R environment for the functions in this package to work. It is highly suggested that
#' you place those in \code{~/.Renviron} at least for interactive work.
#'
#' @param query the SQL query to be executed
#' @param format the format data should be output in. Must be \code{csv} or \code{json}.
#'               Default: \code{csv}.
#' @param flatten should nested and repeated fields in the query results be flattened. Default: \code{true}.
#' @param compress should data files be gzipped. Default: \code{false}.
#' @param delimiter delimiter to use between fields in the exported data. Default: \code{","}.
#' @param headers should a header row be included in results files. Default: \code{true}.
#' @references Censys SQL query syntax: \url{https://censys.io/query};
#'             API doc: \url{https://censys.io/api/v1/docs/export}
#' @return API call result (invisibly)
#' @export
#' @examples \dontrun{
#' q <- censys_start_export("
#' SELECT location.country, count(ip) FROM ipv4.20161206 GROUP BY location.country
#' ")
#' censys_export_job_status(q$job_id)
#' censys_export_download(q$job_id, "~/Data")
#' }
censys_start_export <- function(query, format=c("csv", "json"), flatten=TRUE, compress=FALSE,
                          delimiter=",", headers=TRUE) {

  format <- match.arg(format, c("csv", "json"))

  result <- httr::POST(CENSYS_API_URL %s+% "export",
                       body=list(
                         query=query,
                         format=format,
                         flatten=flatten,
                         compress=compress,
                         delimiter=delimiter,
                         headers=headers),
                       encode="json",
                       check_api())

  httr::stop_for_status(result)

  srs <- jsonlite::fromJSON(content(result, as="text"), flatten=TRUE)

  class(srs) <- c("censys_export_res", class(srs))

  message(sprintf("Query: %s\nStatus: %s\nJob Id: %s",
                  srs$configuration$query, srs$status, srs$job_id))

  invisible(srs)

}

#' Get status of a Censys export job
#'
#' The Get Job Status endpoint lets you retrieved information about a previously submitted
#' job. The status field will return "pending" until the job has completed at which time
#' status will be "success" or "error". On success, the output will define download_paths,
#' a list of files that can be downloaded for the next 24 hours. After 24 hour, the job
#' status will change to "expired" and the files will no longer be retrievable.
#'
#' You must have both \code{CENSYS_API_ID} and \code{CENSYS_API_SECRET} present in the
#' R environment for the functions in this package to work. It is highly suggested that
#' you place those in \code{~/.Renviron} at least for interactive work.
#'
#' @param job_id Censys export job id (from calling \code{censys_start_export()})
#' @references Censys SQL query syntax: \url{https://censys.io/query};
#'             API doc: \url{https://censys.io/api/v1/docs/export}
#' @return API call result (invisibly)
#' @export
#' @examples \dontrun{
#' q <- censys_start_export("
#' SELECT location.country, count(ip) FROM ipv4.20161206 GROUP BY location.country
#' ")
#' censys_export_job_status(q$job_id)
#' censys_export_download(q$job_id, "~/Data")
#' }
censys_export_job_status <- function(job_id) {

  result <- httr::GET(CENSYS_API_URL %s+% "export/" %s+% job_id,
                      check_api())

  httr::stop_for_status(result)

  srs <- jsonlite::fromJSON(content(result, as="text"), flatten=TRUE)

  class(srs) <- c("censys_export_job_status", class(srs))

  message(sprintf("Query: %s\nStatus: %s\nJob Id: %s",
                  srs$configuration$query, srs$status, srs$job_id))

  invisible(srs)

}

#' Download export job files to a specified directory
#'
#' @param job_id Censys export job id (from calling \code{censys_start_export()})
#' @param path Location for downloaded data.
#' @return API call result (invisibly)
#' @export
#' @examples \dontrun{
#' q <- censys_start_export("
#' SELECT location.country, count(ip) FROM ipv4.20161206 GROUP BY location.country
#' ")
#' censys_export_job_status(q$job_id)
#' censys_export_download(q$job_id, "~/Data")
#' }
censys_export_download <- function(job_id, path) {

  path <- path.expand(path)
  if (!dir.exists(path)) stop(sprintf("'%s' does not exist", path), call.=FALSE)

  result <- httr::GET(CENSYS_API_URL %s+% "export/" %s+% job_id,
                      check_api())

  httr::stop_for_status(result)

  srs <- jsonlite::fromJSON(content(result, as="text"), flatten=TRUE)

  if (srs$status != "success") {
    message("Job has not finished yet")
  } else {
    walk(srs$download_paths, function(x) {

      message(sprintf("Downloading %s...", basename(x)))

      fil_path <- file.path(path, basename(x))
      download.file(x, fil_path)

    })

  }

  class(srs) <- c("censys_export_job_results", class(srs))

  invisible(srs)

}


R/report.r
#' Create aggregate reports on the breakdown of a field in the result set of a query
#'
#' The build report endpoint lets you run aggregate reports on the breakdown of
#' a field in a result set analogous to the "Build Report" functionality in the
#' front end. For example, if you wanted to determine the breakdown of cipher
#' suites selected by Top Million Websites.
#'
#' You must have both \code{CENSYS_API_ID} and \code{CENSYS_API_SECRET} present in the
#' R environment for the functions in this package to work. It is highly suggested that
#' you place those in \code{~/.Renviron} at least for interactive work.
#'
#' @param index The search index to be queried. Must be one of either
#'        \code{ipv4}, \code{websites}, or \code{certificates}.
#' @param query The query to be executed. For example,
#'        \code{80.http.get.headers.server: nginx}.
#' @param field The field you are running a breakdown on in "dot notation",
#'        e.g. \code{location.country_code}.
#' @param buckets (optional) The maximum number of values to be returned in the
#'        report. Maximum: \code{500}. Default: \code{50}.
#' @return list of information about the endpoint
#' @references Censys search syntax: \url{https://www.censys.io/ipv4/help};
#'             API doc: \url{https://www.censys.io/api/v1/docs/report}
#' @export
#' @examples \dontrun{
#' censys_report("ipv4", "80.http.get.headers.server: Apache",
#'               "location.country", 100)
#' }
censys_report<- function(index, query, field, buckets=50) {

  result <- httr::POST(CENSYS_API_URL %s+% "report/" %s+% index,
                       body=list(query=query,
                                 field=field,
                                 buckets=buckets),
                       encode="json",
                       check_api())

  httr::stop_for_status(result)

  srs <- jsonlite::fromJSON(httr::content(result, as="text"), flatten=TRUE)

  class(srs) <- c("censys_rpt_res", class(srs))

  srs

}


R/search.r
#' Perform queries against Censys data
#'
#' The search endpoint allows searches against the current Censys data in the
#' IPv4, Top Million Websites, and Certificates indexes using the same search
#' syntax as the primary site. The endpoint returns a paginated result set of
#' hosts (or websites or certificates) that match the search.
#'
#' You must have both \code{CENSYS_API_ID} and \code{CENSYS_API_SECRET} present in the
#' R environment for the functions in this package to work. It is highly suggested that
#' you place those in \code{~/.Renviron} at least for interactive work.
#'
#' @param index The search index to be queried. Must be one of either
#'        \code{ipv4}, \code{websites}, or \code{certificates}.
#' @param query The query to be executed. For example,
#'        \code{80.http.get.headers.server: nginx}.
#' @param page The page of the result set to be returned. The number of pages in
#'        the result set is available under metadata in any request. By default,
#'        the API will return the first page of results. "\code{1}" indexed.
#' @param fields (optional) character vector of fields you would like returned in
#'        the result set in "dot notation", e.g. \code{location.country_code}.
#' @return list of information about the endpoint
#' @references Censys search syntax: \url{https://www.censys.io/ipv4/help};
#'             API doc: \url{https://www.censys.io/api/v1/docs/search}
#' @export
#' @examples \dontrun{
#' censys_search("ipv4", "80.http.get.headers.server: Apache", 2,
#'               c("ip", "location.country", "autonomous_system.asn"))
#' }
censys_search <- function(index=c("ipv4", "websites", "certificates"),
                          query, page=1, fields=NULL) {

  index <- match.arg(index, c("ipv4", "websites", "certificates"))

  result <- httr::POST(CENSYS_API_URL %s+% "search/" %s+% index,
                 body=list(query=query,
                            page=page,
                            fields=fields),
                 encode="json",
                 check_api())

  httr::stop_for_status(result)

  srs <- jsonlite::fromJSON(content(result, as="text"), flatten=TRUE)

  class(srs) <- c("censys_srch_res", class(srs))

  srs

}


R/censys-package.r
#' Tools to access the Censys API
#'
#' Censys is a search engine that allows computer scientists to ask questions
#' about the devices and networks that compose the Internet. Driven by I
#' nternet-wide scanning, Censys lets researchers find specific hosts and
#' create aggregate reports on how devices, websites, and certificates are
#' configured and deployed.
#'
#' The Censys API provides programmatic access to the same data accessible
#' through web interface (\url{https://censys.io/}).
#'
#' You must have both \code{CENSYS_API_ID} and \code{CENSYS_API_SECRET} present in the
#' R environment for the functions in this package to work. It is highly suggested that
#' you place those in \code{~/.Renviron} at least for interactive work.
#'
#' Censys tutorial: \url{https://censys.io/tutorial}
#'
#' @name censys
#' @docType package
#' @author Bob Rudis (brudis@@rapid7.com)
#' @references \url{https://censys.io/about};
#'             \url{https://censys.io/static/censys.pdf}
#' @import httr jsonlite stringi
#' @importFrom purrr walk
#' @importFrom utils download.file
NULL


R/sqlquery.r
#' Issue SQL Queries against the Censys API
#'
#' The Query API allows executing SQL queries against our daily snapshots and raw data
#' analogous to the Query web interface. Queries are executed asynchronously. You must
#' first start a job, then check its status. Once a job has completed, you can view
#' paginated results using the get results endpoint. Jobs typically require 15-30
#' seconds to execute; results can be viewed for 24 hours after the job completed.
#' Ddefinition endpoints are also exposed where you can list the series and view series
#' details (i.e., list tables and schema).
#'
#' You must have both \code{CENSYS_API_ID} and \code{CENSYS_API_SECRET} present in the
#' R environment for the functions in this package to work. It is highly suggested that
#' you place those in \code{~/.Renviron} at least for interactive work.
#'
#' @param sql SQL query string
#' @references Censys SQL query syntax: \url{https://censys.io/query};
#'             API doc: \url{https://www.censys.io/api/v1/docs/search}
#' @export
#' @return API call result (invisibly)
#' @examples \dontrun{
#' q <- censys_query("SELECT p443.https.tls.cipher_suite.name, count(ip) FROM ipv4
#'                    WHERE p443.https.tls.validation.browser_trusted=true
#'                    GROUP BY p443.https.tls.cipher_suite.name;")
#' censys_get_job_status(q$job_id)
#' censys_get_job_result(q$job_id)
#' }
censys_query <- function(sql) {

  result <- httr::POST(CENSYS_API_URL %s+% "query",
                       body=list(query=sql),
                       encode="json",
                       check_api())

  httr::stop_for_status(result)

  srs <- jsonlite::fromJSON(content(result, as="text"), flatten=TRUE)

  class(srs) <- c("censys_query_res", class(srs))

  message(sprintf("Query: %s\nStatus: %s\nJob Id: %s",
                  srs$configuration$query, srs$status, srs$job_id))

  invisible(srs)

}


#' Get status of a Censys SQL query job
#'
#' The Get Job Status endpoint allows you to determine whether a job has completed. Once
#' it has successfully finished, you can then retrieved results with the Get Results
#' endpoint. Data should be posted as a JSON request document.
#'
#' You must have both \code{CENSYS_API_ID} and \code{CENSYS_API_SECRET} present in the
#' R environment for the functions in this package to work. It is highly suggested that
#' you place those in \code{~/.Renviron} at least for interactive work.
#'
#' @param job_id Censys job id (from calling \code{censys_query()})
#' @references Censys SQL query syntax: \url{https://censys.io/query};
#'             API doc: \url{https://www.censys.io/api/v1/docs/search}
#' @return API call result (invisibly)
#' @export
#' @examples \dontrun{
#' q <- censys_query("SELECT p443.https.tls.cipher_suite.name, count(ip) FROM ipv4
#'                    WHERE p443.https.tls.validation.browser_trusted=true
#'                    GROUP BY p443.https.tls.cipher_suite.name;")
#' censys_get_job_status(q$job_id)
#' censys_get_job_result(q$job_id)
#' }
censys_get_job_status <- function(job_id) {

  result <- httr::GET(CENSYS_API_URL %s+% "query/" %s+% job_id,
                      check_api())

  httr::stop_for_status(result)

  srs <- jsonlite::fromJSON(content(result, as="text"), flatten=TRUE)

  class(srs) <- c("censys_job_status", class(srs))

  message(sprintf("Query: %s\nStatus: %s\nJob Id: %s",
                  srs$configuration$query, srs$status, srs$job_id))

  invisible(srs)

}

#' Get results of completed Censys SQL query job
#'
#' The Get Results endpoint allows you to retrieve results of a query after it has
#' completed successfully.
#'
#' You must have both \code{CENSYS_API_ID} and \code{CENSYS_API_SECRET} present in the
#' R environment for the functions in this package to work. It is highly suggested that
#' you place those in \code{~/.Renviron} at least for interactive work.
#'

#' @param job_id Censys job id (from calling \code{censys_query()})
#' @param page page number of paged results
#' @references Censys SQL query syntax: \url{https://censys.io/query};
#'             API doc: \url{https://www.censys.io/api/v1/docs/search}
#' @export
#' @examples \dontrun{
#' q <- censys_query("SELECT p443.https.tls.cipher_suite.name, count(ip) FROM ipv4
#'                    WHERE p443.https.tls.validation.browser_trusted=true
#'                    GROUP BY p443.https.tls.cipher_suite.name;")
#' censys_get_job_status(q$job_id)
#' censys_get_job_result(q$job_id)
#' }
censys_get_job_result <- function(job_id, page=1) {

  result <- httr::GET(CENSYS_API_URL %s+% "query/" %s+% job_id %s+% "/" %s+% page,
                      check_api())

  httr::stop_for_status(result)

  srs <- jsonlite::fromJSON(content(result, as="text"), flatten=TRUE)

  class(srs) <- c("censys_job_results", class(srs))

  srs

}

#' List all series that can be queried from the SQL interface
#'
#' You must have both \code{CENSYS_API_ID} and \code{CENSYS_API_SECRET} present in the
#' R environment for the functions in this package to work. It is highly suggested that
#' you place those in \code{~/.Renviron} at least for interactive work.
#'
#' @references Censys SQL query syntax: \url{https://censys.io/query};
#'             API doc: \url{https://www.censys.io/api/v1/docs/search}
#' @export
#' @examples \dontrun{
#' censys_series()
#' }
censys_series <- function() {

  result <- httr::GET(CENSYS_API_URL %s+% "query_definitions",
                      check_api())

  httr::stop_for_status(result)

  srs <- jsonlite::fromJSON(content(result, as="text"), flatten=TRUE)

  class(srs) <- c("censys_series", class(srs))

  srs

}

#' Get details about a series, including the list of tables and schema for the series
#'
#' You must have both \code{CENSYS_API_ID} and \code{CENSYS_API_SECRET} present in the
#' R environment for the functions in this package to work. It is highly suggested that
#' you place those in \code{~/.Renviron} at least for interactive work.
#'
#' @param series series name (call \code{censys_series() to see them all}). Defaults to \code{ipv4}.
#' @references Censys SQL query syntax: \url{https://censys.io/query};
#'             API doc: \url{https://www.censys.io/api/v1/docs/search}
#' @export
#' @examples \dontrun{
#' censys_series_details("ipv4")
#' }
censys_series_details <- function(series="ipv4") {

  result <- httr::GET(CENSYS_API_URL %s+% "query_definitions/" %s+% series,
                      check_api())

  httr::stop_for_status(result)

  srs <- jsonlite::fromJSON(content(result, as="text"), flatten=TRUE)

  class(srs) <- c("censys_series_details", class(srs))

  srs

}
